internal/controller/dnsrecord_healthchecks.go

```diff
@@ -0,0 +1,242 @@
+package controller
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/net"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/provider"
+)
+
+// healthChecksConfig represents the user configuration for the health checks
+type healthChecksConfig struct {
+	Endpoint         string
+	Port             *int64
+	FailureThreshold *int64
+	Protocol         *provider.HealthCheckProtocol
+}
+
+func (r *DNSRecordReconciler) ReconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord) error {
+	var results []provider.HealthCheckResult
+	var err error
+
+	// Get the configuration for the health checks. If no configuration is
+	// set, ensure that the health checks are deleted
+	config := getHealthChecksConfig(dnsRecord)
+
+	log.Log.Info("DNS Record healthcheck status", "status", dnsRecord.Status.HealthCheck)
+	if config == nil {
+		log.Log.V(3).Info("No health check for DNSPolicy, ensuring deletion")
+		results, err = r.reconcileHealthCheckDeletion(ctx, dnsRecord)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Log.V(3).Info("reconciling health checks")
+		results, err = r.reconcileHealthChecks(ctx, dnsRecord, config)
+		if err != nil {
+			return err
+		}
+	}
+
+	result := r.reconcileHealthCheckStatus(results, dnsRecord)
+	log.Log.Info("health check status reconcile", "result", result)
+	return result
+}
+
+func (r *DNSRecordReconciler) reconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord, config *healthChecksConfig) ([]provider.HealthCheckResult, error) {
+	dnsProvider, err := r.getDNSProvider(ctx, dnsRecord)
+	if err != nil {
+		return nil, err
+	}
+	healthCheckReconciler := dnsProvider.HealthCheckReconciler()
+
+	var results []provider.HealthCheckResult
+	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
+		log.Log.Info("reconciling DNS Record health checks", "endpoint", dnsEndpoint)
+		address, ok := provider.GetEndpointAddress(dnsEndpoint)
+		if !ok {
+			log.Log.Info("address not ok", "address", address)
+			continue
+		} else if !net.IsIPv4String(address) {
```

Comment:
intresting so will we not have any health checks if the using an AWS LB as it uses hostnames?

Reply:
ah good point, I need to also capture when it's an external CNAME.